addons/sale/models/analytic.py

```diff
@@ -0,0 +1,145 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = "account.analytic.line"
+
+    so_line = fields.Many2one('sale.order.line', string='Sales Order Line')
+
+    @api.model
+    def create(self, values):
+        result = super(AccountAnalyticLine, self).create(values)
+        result._sale_postprocess()
+        return result
+
+    @api.multi
+    def write(self, values):
+        # get current so lines for which update qty wil be required
+        sale_order_lines = self.env['sale.order.line']
+        if 'so_line' in values:
+            sale_order_lines = self.sudo().mapped('so_line')
+        result = super(AccountAnalyticLine, self).write(values)
+        self._sale_postprocess(values, force_so_lines=sale_order_lines)
+        return result
+
+    @api.multi
+    def unlink(self):
+        sale_order_lines = self.sudo().mapped('so_line')
+        res = super(AccountAnalyticLine, self).unlink()
+        self.env['account.analytic.line']._sale_postprocess(force_so_lines=sale_order_lines)
+        return res
+
+    @api.multi
+    def _sale_postprocess(self, values=None, force_so_lines=None):
+        values = values or {}
+
+        if 'so_line' not in values:  # allow to force a False value for so_line
+            self.with_context(sale_analytic_norecompute=True)._sale_determine_order_line()
+
+        if any(field_name in values for field_name in ['so_line', 'unit_amount', 'product_uom_id']):
```

Comment:
Suggestion: move those key fields in a method to be able to modify / inherit it easily ?